2.2 release notes plus some minor doc changes

roji opened the pull request:
Please see these proposed release notes for 2.2. The idea is to keep it to a relative minimum, and only mention impactful, user-facing items. I've also done a bit of work on index modeling.

I think our docs could use some work - many SQL translations are undocumented, and the "misc" page looks a bit like an unsorted bin... But I propose to do that after the 2.2 release.

Let's try to finish this up today so I can release either today or tomorrow morning!

/cc @YohDeadfall in case you want to look (I know you're busy).

austindrenski commented on doc/modeling/indices.md:
```diff
@@ -0,0 +1,42 @@
+# Indices
+
+## PostgreSQL covering indices (INCLUDE)
```
I've mostly seen the "indexes" version written out. Any reason to use indices here?

roji replied:
My grammar-obsessive Latin side is what makes me use indices. It's true that indexes is more common, will change it everywhere.

austindrenski commented on doc/modeling/indices.md:
```diff
@@ -0,0 +1,42 @@
+# Indices
```
As below

austindrenski commented on doc/modeling/indices.md:
````diff
@@ -0,0 +1,42 @@
+```
+
+This will create an index for searching on `Id`, but containing also the column `Name`, so that reading the latter will not involve accessing the table. The SQL generated is as follows:
+
````
Can we add the code sample before merging?

roji replied:
Oops, I left that one out - fixed.

austindrenski commented on doc/modeling/indices.md:
```diff
@@ -0,0 +1,42 @@
+
+PostgreSQL supports a number of _index methods_, or _types_. These are specified at index creation time via the `USING _method_` clause, see the [PostgreSQL docs for `CREATE INDEX`](https://www.postgresql.org/docs/current/static/sql-createindex.html) and [this page](https://www.postgresql.org/docs/current/static/indexes-types.html) for information on the different types.
+
+The Npgsql EF Core provider allows you to specify the index method to be used by specifying `ForNpgsqlHasMethod()` on your index in your context's `OnModelCreating` method:
```
"to be used by calling" (less repetitive)

austindrenski commented on doc/modeling/indices.md:
````diff
@@ -0,0 +1,42 @@
+```c#
+protected override void OnModelCreating(ModelBuilder modelBuilder)
+{
+    => modelBuilder.Entity<Blog>()
````
Looks like a copy/paste error here with the method body syntax.

austindrenski commented on doc/modeling/indices.md:
````diff
@@ -0,0 +1,42 @@
+
+```c#
+protected override void OnConfiguring(DbContextOptionsBuilder builder)
+{
````
Little nit, but I'd use expression body syntax throughout where possible.

roji replied:
Agreed!

austindrenski commented on doc/modeling/indices.md:
```diff
@@ -0,0 +1,42 @@
+
+## PostgreSQL index operator classes
+
+PostgreSQL allows you to specify [operator classes on your indices](https://www.postgresql.org/docs/current/indexes-opclass.html), to allow tweaking how the index should work. Use the following code to specify an operator class:
```
nit: consistency for indices or indexes

austindrenski commented on doc/modeling/indices.md:
````diff
@@ -0,0 +1,42 @@
+    .ForNpgsqlHasOperators(null, "text_pattern_ops");
+```
+
+Note that each operator class is used for the corresponding index column, by order. In the example above, the `text_pattern_ops` class will be used for the `Name` column, whereas the `Id` column will have the default class (unspecified)
````
Little nit, but you could swap "whereas" with "while" for sentence-flow.

Also swap "will have the" with "will use the" for comparability to the first clause.

roji replied:
You're the native speaker :)

austindrenski commented on doc/release-notes/2.2.md:
```diff
@@ -1,10 +1,65 @@
+
+This will create an index for searching on `Id`, but containing also the column `Name`, so that reading the latter will not involve accessing the table. [See the documentation for more details](../modeling/indices.md).
+
+Thanks to @khellang for contributing this!
```
Can you use `[@khellang](.....)` so this links correctly to his profile when published in docfx? 

(Not with code ticks, those are just here for syntax clarity)

austindrenski commented on doc/release-notes/2.2.md:
````diff
@@ -1,10 +1,65 @@
+```c#
+protected override void OnConfiguring(DbContextOptionsBuilder builder) {
+    builder.UseNpgsql("...", b => b.MapRange("floatrange", typeof(float)));
+}
````
nit: expression-bodied

austindrenski replied:
Also, isn't there a generic overload for `MapRange`? If so, let's encourage that version here.

roji replied:
Done

austindrenski commented on doc/release-notes/2.2.md:
```diff
@@ -1,10 +1,65 @@
+
+protected override void OnModelCreating(ModelBuilder builder) {
+    builder.ForNpgsqlHasRange("floatrange", "real");
+}
```
As above

austindrenski commented on doc/release-notes/2.2.md:
```diff
@@ -1,10 +1,65 @@
+
+### Type-constrained PostGIS columns
+
+The NetTopologySuite previously created PostGIS columns of type `GEOMETRY` (or `GEOGRAPHY`) in the database. With ([#719](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/issues/719)), a property of type `Point` will be instead mapped to columns of type `GEOMETRY (POINT)`.
```
I think you can drop the parentheses around the link, no?

roji replied:
Yep, some copying-pasting from a previous version...

austindrenski commented on doc/release-notes/2.2.md:
```diff
@@ -1,10 +1,65 @@
+
+### Seeding for Npgsql-specific types
+
+When using some Npgsql-specific types, it wasn't possible to seed values for those types. With EF Core support for seeding any type, ([#667](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/issues/667)) allows seeding values for network, bit and range types (more are coming).
```
As above, the extra parentheses look out of place.

austindrenski commented on doc/release-notes/2.2.md:
```diff
@@ -1,10 +1,65 @@
 # 2.2 Release Notes
 
+Version 2.2.0 of the Npgsql Entity Framework Core provider has been released and is available on nuget. This version works with [version 2.2.0 of Entity Framework Core](https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-2.2), and contains some new Npgsql features as well.
```
I thought it used to be possible to omit "en-us" from the docs links so that the user's locale/region would be used to redirect them to their local language docs.

Though of course, _our_ docs are English-only....but something to consider.

roji replied:
Good idea! Done here and everywhere else.

austindrenski commented on doc/release-notes/2.2.md:
```diff
@@ -1,10 +1,65 @@
+* Various issues with enum and range types were fixed, including upper/lower case, quoting and schema management.
+* Many new SQL translations were added, so more of your LINQ expressions can run in the database. We'll be working on our documentation to make these more discoverable.
 
+The full list of issues for this release is [available here](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/issues?page=2&q=milestone%3A2.2.0+is%3Aclosed&utf8=%E2%9C%93).
```
Is that link to page 2 of the issues list? (Reviewing on mobile, so I can't click it at the moment.)

roji replied:
Great catch, thanks!

austindrenski commented on doc/toc.md:
```diff
@@ -4,6 +4,8 @@
 ## [2.0](release-notes/2.0.md)
 ## [1.1](release-notes/1.1.md)
+# Creating a Model
+## [Indices](modeling/indices.md)
```
As above with the naming. 

To clarify, either would be fine with me, as long as we're consistent and following how PostgreSQL uses these word-versions.

roji replied:
Great catch, thanks...

roji commented:
@austindrenski thanks for the meticulous review! I think I've addressed everything, please take another look.

Note that I also changed some of the docs on user-defined ranges (in range.md), replacing the tabs which you had placed there (2.1 vs 2.2) with a simple informational note on the feature being introduced in 2.2.

austindrenski commented:
> Note that I also changed some of the docs on user-defined ranges (in range.md), replacing the tabs which you had placed there (2.1 vs 2.2) with a simple informational note on the feature being introduced in 2.2.

That's fine... The tabs worked nicely, but the lack of rendering support from GitHub/Rider turned out to be a pain. Given our challenges keeping the docs synchronized tabs may turn out to not be the best solution after all.

roji commented:
I do like the tabs a lot! I just don't really want us to maintain parallel documentation on several versions, I don't think it's justified - even EF Core themselves only put a "new since version X" note.
